network/connecteventmanager.go

```diff
@@ -11,96 +11,198 @@ type ConnectionListener interface {
 }
 
-func (c *connectEventManager) Connected(p peer.ID) {
+func (c *connectEventManager) Start() {
+	go c.worker()
+}
+
+func (c *connectEventManager) Stop() {
 	c.lk.Lock()
-	defer c.lk.Unlock()
+	c.stop = true
+	c.lk.Unlock()
+	c.cond.Broadcast()
+
+	<-c.done
+}
+
+func (c *connectEventManager) getState(p peer.ID) state {
+	if state, ok := c.peers[p]; ok {
+		return state.newState
+	} else {
+		return stateDisconnected
+	}
+}
 
-	state, ok := c.conns[p]
+func (c *connectEventManager) setState(p peer.ID, newState state) {
+	state, ok := c.peers[p]
 	if !ok {
-		state = &connState{responsive: true}
-		c.conns[p] = state
+		state = new(peerState)
+		c.peers[p] = state
+	}
+	state.newState = newState
+	if !state.pending && state.newState != state.curState {
+		state.pending = true
+		c.changeQueue = append(c.changeQueue, p)
+		c.cond.Broadcast()
+	}
+}
+
+func (c *connectEventManager) worker() {
+	c.lk.Lock()
+	defer c.lk.Unlock()
+	defer close(c.done)
+
+	for {
+		for !c.stop && len(c.changeQueue) == 0 {
+			c.cond.Wait()
+		}
+
+		if c.stop {
+			return
+		}
+
+		pid := c.changeQueue[0]
+		c.changeQueue[0] = peer.ID("")
+		c.changeQueue = c.changeQueue[1:]
+
+		state, ok := c.peers[pid]
+		// If we've disconnected and forgotten, continue. We shouldn't reach this?
+		if !ok {
+			continue
+		}
+
+		// Record the fact that this "state" is no longer in the queue.
+		state.pending = false
+
+		// Then, if there's nothing to do, continue.
+		if state.curState == state.newState {
+			continue
+		}
+
+		// Or record the state update, then apply it.
+		oldState := state.curState
+		state.curState = state.newState
+
+		switch state.newState {
+		case stateDisconnected:
+			delete(c.peers, pid)
+			fallthrough
+		case stateUnresponsive:
+			// Only trigger a disconnect event if the peer was responsive.
+			// We could be transitioning from unresponsive to disconnected.
+			if oldState == stateResponsive {
+				c.lk.Unlock()
+				c.connListener.PeerDisconnected(pid)
+				c.lk.Lock()
+			}
+		case stateResponsive:
+			c.lk.Unlock()
+			c.connListener.PeerConnected(pid)
+			c.lk.Lock()
+		}
 	}
-	state.refs++
+}
+
+// Called whenever we receive a new connection. May be called many times.
+func (c *connectEventManager) Connected(p peer.ID) {
+	c.lk.Lock()
+	defer c.lk.Unlock()
 
-	if state.refs == 1 && state.responsive {
-		c.connListener.PeerConnected(p)
+	// disconnected -> responsive
+
+	if c.getState(p) != stateDisconnected {
+		return
 	}
+	c.setState(p, stateResponsive)
 }
 
+// Called when we drop the final connection to a peer.
 func (c *connectEventManager) Disconnected(p peer.ID) {
 	c.lk.Lock()
 	defer c.lk.Unlock()
 
-	state, ok := c.conns[p]
-	if !ok {
-		// Should never happen
+	// !disconnected -> disconnected
+
+	if c.getState(p) == stateDisconnected {
 		return
 	}
-	state.refs--
 
-	if state.refs == 0 {
-		if state.responsive {
-			c.connListener.PeerDisconnected(p)
-		}
-		delete(c.conns, p)
-	}
+	c.setState(p, stateDisconnected)
```

Comment:
This will end up reading the state map _twice_. That's not great, but it simplifies the logic quite a bit (especially around enqueuing changes).

An alternative would be a `c.transition(condiitonFn, newState)` function.